Support DELETE statement with partition key for Cassandra

t2y opened the pull request:
# Description

## Why

Cassandra Connector doesn't support DELETE statement at all.

This PR adds to support DELETE statement with a given partition key only. Deleting with the partition key is not useful for production use. It can be done with [cqlsh](https://docs.datastax.com/en/cql-oss/3.x/cql/cql_reference/cqlsh.html). So I expect this feature is for development/testing use.

However, this is a first step to support DELETE statement for Cassandra Connector. For production use, I want to add a select-delete feature divides split tasks using `TOKEN` function in Cassandra. But, I know it has another issue though I won't mention here.

## What

I added these interface methods for CassandraMetadata.

* getUpdateRowIdColumnHandle()
* applyDelete()
* executeDelete()

~~To be able to delete data, we set `cassandra.allow-delete-data` property.~~

# Reference

* https://prestosql.io/docs/current/connector/cassandra.html

t2y commented:
@cla-bot check

ebyhr commented on presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/util/CassandraCqlUtils.java:
```diff
@@ -121,4 +129,27 @@ public static Select selectDistinctFrom(CassandraTableHandle tableHandle, List<C
+        Optional<List<CassandraPartition>> opt = handle.getPartitions();
+        List<CassandraPartition> partitions = opt.orElse(Collections.emptyList());
+        String clause = getWhereClause(partitions, handle.getClusteringKeyPredicates());
+        String cql = delete.getQueryString().replace(";", clause);
```
If we can use Cassandra's `QueryBuilder` to build clause, it worth to use the library. But `StringBuilder` looks sufficient and then we can avoid this hack.

t2y replied:
I see. I will construct CQL with `StringBuilder` only.

ebyhr commented on presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/util/CassandraCqlUtils.java:
```diff
@@ -121,4 +129,27 @@ public static Select selectDistinctFrom(CassandraTableHandle tableHandle, List<C
+        List<CassandraPartition> partitions = opt.orElse(Collections.emptyList());
+        String clause = getWhereClause(partitions, handle.getClusteringKeyPredicates());
+        String cql = delete.getQueryString().replace(";", clause);
+        log.debug("Created CQL: %s", cql);
```
Please move logging to caller side to focus the actual logic in util class.

t2y replied:
Got it.

ebyhr commented on presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/util/CassandraCqlUtils.java:
```diff
@@ -121,4 +129,27 @@ public static Select selectDistinctFrom(CassandraTableHandle tableHandle, List<C
         return from(select(columns).distinct(), tableHandle);
     }
+
+    public static String getWhereClause(List<CassandraPartition> partitions, String clusteringKeyPredicates)
```
No need to extract this method because it's simple and only used in `deleteFrom`.

t2y replied:
OK. Got it.

t2y commented on presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraClientConfig.java:
```diff
@@ -189,6 +190,19 @@ public CassandraClientConfig setAllowDropTable(boolean allowDropTable)
+        return this.allowDeleteData;
+    }
+
+    @Config("cassandra.allow-delete-data")
```
No, I don't need it in my environment. I followed `@Config("cassandra.allow-drop-table")` and I just thought deleting data is more careful. But, I will remove it if no need.

t2y commented on presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java:
```diff
@@ -369,4 +373,41 @@ private static boolean isHiddenIdColumn(CassandraColumnHandle columnHandle)
+    public OptionalLong executeDelete(ConnectorSession session, ConnectorTableHandle deleteHandle)
+    {
+        CassandraTableHandle handle = (CassandraTableHandle) deleteHandle;
+        Optional<List<CassandraPartition>> opt = handle.getPartitions();
+        if (!opt.isPresent()) {
```
I see. Got it.

t2y commented on presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java:
```diff
@@ -369,4 +373,41 @@ private static boolean isHiddenIdColumn(CassandraColumnHandle columnHandle)
+            throw new PrestoException(PERMISSION_DENIED, "Deleting without partition key is unsupported");
+        }
+        if (opt.get().isEmpty()) {
+            throw new PrestoException(PERMISSION_DENIED, "No data with the given condition");
```
When a partition does not exist for the specified condition, we shouldn't execute a delete statement. At that time, should I return just `OptionalLong.empty();`? I'm not sure what handling is better for that condition.

t2y replied:
I thought again about the below code.

```java
if (partitions.get().isEmpty()) {
    throw new PrestoException(PERMISSION_DENIED, "No data with the given condition");
}
```

> I suppose the query should not be failed

Probably true. Cassandra cluster will return an error when it requests a delete statement without partition keys if the partitions are empty. So, I don't care about it.

I fixed it in 1b9db106.

t2y replied:
I had forgotten that code prevents an error when there are no records with a given partition key.

```
Query 20200626_042205_00023_qzyzn failed: line 0:-1 no viable alternative at input '<EOF>'
```

I changed to return without any error in acceec1b.

t2y commented:
> Could you add test to TestCassandraIntegrationSmokeTest and update TestCassandraDistributedQueries.testDelete?

I have some questions.

For TestCassandraIntegrationSmokeTest, is this smoke test to confirm Cassandra cluster's behavior? I'm not sure what purpose is this smoke test?

For TestCassandraDistributedQueries, it seems this test suit is not for Cassandra and some insert tests are TODO. Can I update as a TODO comment like this?

```java
    @Override
    public void testDelete()
    {
        // TODO test delete
    }
```

t2y commented:
I rebased from `master` to pass `presto-druid` tests.

ebyhr commented:
@t2y The purpose is testing by query level. The current test you added is calling each methods directly, but we would like to test with `DELETE` statement in Presto. 

Could you update testDelete like this? Thus, we can aware the future change by CI failure. 
```java
        assertThatThrownBy(super::testDelete)
                .hasStackTraceContaining("This connector only supports delete with primary key or partition key");
```

t2y commented:
@ebyhr Thank you for your reply. I learnt what purpose TestCassandraIntegrationSmokeTest is. I will implement DELETE statement tests.

findepi commented:
The discussion here prompted me to clarify -- https://github.com/prestosql/presto/pull/4132 🙂

kokosing commented on presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java:
```diff
@@ -369,4 +374,39 @@ private static boolean isHiddenIdColumn(CassandraColumnHandle columnHandle)
+    @Override
+    public ColumnHandle getUpdateRowIdColumnHandle(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        return new CassandraColumnHandle("$dummy", 0, CassandraType.TEXT, false, false, false, true);
```
s/dummy/update_row_id

kokosing commented on presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java:
```diff
@@ -369,4 +374,39 @@ private static boolean isHiddenIdColumn(CassandraColumnHandle columnHandle)
+        CassandraTableHandle handle = (CassandraTableHandle) deleteHandle;
+        Optional<List<CassandraPartition>> partitions = handle.getPartitions();
+        if (!partitions.isPresent()) {
+            throw new PrestoException(PERMISSION_DENIED, "Deleting without partition key is unsupported");
```
s/PERMISSION_DENIED/NOT_SUPPORTED ?

kokosing commented on presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java:
```diff
@@ -369,4 +374,39 @@ private static boolean isHiddenIdColumn(CassandraColumnHandle columnHandle)
+            throw new PrestoException(PERMISSION_DENIED, "Deleting without partition key is unsupported");
+        }
+        if (partitions.get().isEmpty()) {
+            throw new PrestoException(PERMISSION_DENIED, "No data with the given condition");
```
s/PERMISSION_DENIED/NOT_SUPPORTED ?

kokosing commented on presto-cassandra/src/main/java/io/prestosql/plugin/cassandra/CassandraMetadata.java:
```diff
@@ -369,4 +374,39 @@ private static boolean isHiddenIdColumn(CassandraColumnHandle columnHandle)
+            throw new PrestoException(PERMISSION_DENIED, "No data with the given condition");
+        }
+        String cql = CassandraCqlUtils.deleteFrom(handle);
+        log.debug("Delete statement: %s", cql);
```
maybe logging should happen in `CassandraSession`?

t2y commented:
I almost fixed code from review comments. Tell me if you have other concerns.

ebyhr commented:
Could you confirm the following behavior? It would be better to throw more friendly exception message. 
```sql
presto:tpch> delete from cassandra.tpch.test where id = 'e138d065-6357-4f04-8490-9be29ad7a988';
DELETE

presto:tpch> delete from cassandra.tpch.test where id = 'e138d065-6357-4f04-8490-9be29ad7a988';
Query 20200625_140704_00015_3mq56 failed: line 0:-1 no viable alternative at input '<EOF>'
```

t2y commented:
Hmm, I didn't know that error. I will investigate.

t2y commented:
I could reproduce. It occurred with partition key when there are no records of the partition key. I will fix it.

```
presto:test> delete from mytable where insight_id=2 and chunk_num=32 and id=22;
Query 20200626_042158_00022_qzyzn failed: Some partition key parts are missing: insight_id, chunk_num

presto:test> delete from mytable where insight_id=2 and chunk_num=32;
Query 20200626_042205_00023_qzyzn failed: line 0:-1 no viable alternative at input '<EOF>'
```

t2y commented:
I rebased from `master` branch.

ebyhr commented on presto-cassandra/src/test/java/io/prestosql/plugin/cassandra/TestCassandraConnector.java:
```diff
@@ -267,4 +272,48 @@ private ConnectorTableHandle getTableHandle(SchemaTableName tableName)
+    public void testExecuteDelete()
+    {
+        assertNumberOfRows(tableForDelete, 15);
+        CassandraTableHandle handle1 = getTableHandle(Optional.of(
+                List.of(createPartition(1, 1))), "");
```
nit: No need to wrap. Following block is also the same.

t2y commented on presto-cassandra/src/test/java/io/prestosql/plugin/cassandra/CassandraTestingUtils.java:
```diff
@@ -248,4 +250,62 @@ private static void insertTestData(CassandraSession session, SchemaTableName tab
+                "partition_two int, " +
+                "clust_one text, " +
+                "data text, " +
+                "PRIMARY KEY((partition_one, partition_two), clust_one) " + ")");
```
Looking around other test code, the format is like this.

```java
session.execute("CREATE TABLE " + table + " (" +
        ...
        "data text, " +
        "PRIMARY KEY((partition_one, partition_two), clust_one, clust_two, clust_three) " +
        ")");
```

t2y commented:
@ebyhr Thank you for your review. I fixed some you commented.

> Please add a test case which tries to delete several partitions at once. Current implementation gets exception from Cassandra due to the wrong cql condition (e.g. WHERE id = 1 AND id = 2). Supporting deletion only one partition is no problem as the first step, but it would be better to throw an exception before executing cql.

You are right. I don't know how to validate the cql condition before executing cql. Do you have good advice on how to implement to validate it?

ebyhr commented:
@t2y Could you try calling multiple `DELETE` cql to resolve it as follows? I suppose it's easier than validating cql. 
```sql
DELETE FROM x.y WHERE id = 1;
DELETE FROM x.y WHERE id = 2;
```

t2y commented:
@ebyhr Thank you for your advice. I could understand some cases the presto handles multiple partitions. I fixed it in db858dd2 to execute multiple delete statements and added some tests.

ebyhr reviewed (approved):
LGTM except for minor comments. Please squash commits.

t2y commented on presto-cassandra/src/test/java/io/prestosql/plugin/cassandra/TestCassandraIntegrationSmokeTest.java:
```diff
@@ -629,6 +631,43 @@ public void testInsert()
+        String keyspaceAndTable = format("%s.%s", KEYSPACE, TABLE_DELETE_DATA);
+        assertEquals(execute("SELECT * FROM " + keyspaceAndTable).getRowCount(), 15);
+
+        // error
```
Because there are deleting row tests and the successful test cases delete test data for error cases. So, I do tests for error cases first.

t2y commented:
@ebyhr Thank you for your review. I squashed into two commits.

ebyhr commented:
@t2y Could you rebase on upstream to resolve a conflict? I will merge after the operation.

t2y commented:
@ebyhr Could you merge this? I won't maintain this any longer.

ebyhr commented:
Merged, thanks! Sorry, I was missing your last push. Please feel free to ping after pushing commits from next time.
